### PR TITLE
Add type annotation logic for complex numbers

### PR DESCRIPTION
**Pull request by anjali411**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51145 Support complex number list in JIT
* #50885 Add serialization logic for complex numbers
* **#50884 Add type annotation logic for complex numbers**

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D26086963](https://our.internmc.facebook.com/intern/diff/D26086963)

### CODE REVIEW

File: test/test_complex.py [hunk 1]
```diff
@@ -1,9 +1,22 @@
 import torch
 from torch.testing._internal.common_device_type import instantiate_device_type_tests, dtypes
 from torch.testing._internal.common_utils import TestCase, run_tests
+from torch.testing._internal.jit_utils import JitTestCase
 
 devices = (torch.device('cpu'), torch.device('cuda:0'))
 
+class TestComplexJIT(JitTestCase):
```
Comment (SplitInfinity) [inline test/test_complex.py:8]:
```suggestion
class TestComplex(JitTestCase):
```

File: test/test_complex.py [hunk 2]
```diff
@@ -1,9 +1,22 @@
...
+    def test_script(self):
+        def fn(a: complex):
+            return a
+
+        b = 3 + 5j
+        result = self.checkScript(fn, (b,))
+        cu = torch.jit.CompilationUnit('''
+                def fn(a: complex):
+                    return a
+        ''')
```
Comment (SplitInfinity) [inline test/test_complex.py:18]: `checkScript` should test this internally as well, so I think you can delete this.

File: torch/csrc/jit/python/pybind_utils.cpp [hunk 3]
```diff
@@ -229,6 +233,9 @@ IValue toIValue(py::handle obj, const TypePtr& type, c10::optional<int32_t> N) {
         return py::cast<int64_t>(obj);
       } else if (py::isinstance<py::float_>(obj)) {
         return py::cast<double>(obj);
+      } else if (PyComplex_CheckExact(obj.ptr())) {
```
Comment (SplitInfinity) [inline torch/csrc/jit/python/pybind_utils.cpp:236]: Dumb question but why does this line not say something like `py::isinstance<py::complex_>(...)` like all the others?

File: torch/jit/annotations.py [hunk 4]
```diff
@@ -305,6 +305,8 @@ def try_ann_to_type(ann, loc):
         return FutureType(try_ann_to_type(ann.__args__[0], loc))
     if ann is float:
         return FloatType.get()
+    if inspect.isclass(ann) and issubclass(ann, builtins.complex):
+        return ComplexDoubleType.get()
```
Comment (SplitInfinity) [inline torch/jit/annotations.py:309]: Another dumb question, why is this not `ann is complex`?

Comment (SplitInfinity) [reply torch/jit/annotations.py:309]: Also, is there a usecase for creating subclasses of `complex`?

File: torch/csrc/jit/frontend/function_schema_parser.cpp [hunk 5]
```diff
@@ -248,6 +248,7 @@ struct SchemaParser {
       case TypeKind::IntType:
       case TypeKind::BoolType:
       case TypeKind::FloatType:
+      case TypeKind::ComplexDoubleType:
         return parseSingleConstant(arg_type->kind());
```
Comment (SplitInfinity) [inline torch/csrc/jit/frontend/function_schema_parser.cpp:252]: `parseSingleConstant` needs to be updated to handle complex numbers. I would either get rid of it and do it in a follow-up PR or add a test for it (and update `parseSingleConstant` and other code as needed).

**Review by SplitInfinity — COMMENTED** [review]
Nice! Do you plan to support constant literals in TorchScript code in this diff? If not, I think it is ready to land. If so, I think some more changes are required.

File: torch/csrc/jit/python/pybind_utils.cpp (hunk 3 above)
Comment (anjali411) [reply torch/csrc/jit/python/pybind_utils.cpp:236]: Complex type is not available in pybind (see doc here: https://pybind11.readthedocs.io/en/stable/advanced/pycpp/object.html#accessing-python-libraries-from-c) so I just pass the underlying `PyObject*` to `PyComplex_CheckExact` to check whether it's complex

File: torch/jit/annotations.py (hunk 4 above)
Comment (anjali411) [reply torch/jit/annotations.py:309]: complex is a python class hence the `isclass` call. umm no there's no use specific case of creating subclasses of complex

File: torch/csrc/jit/frontend/function_schema_parser.cpp (hunk 5 above)
Comment (anjali411) [reply torch/csrc/jit/frontend/function_schema_parser.cpp:252]: I'll just remove this change for the current PR and add it separately

File: torch/jit/annotations.py (hunk 4 above)
Comment (SplitInfinity) [reply torch/jit/annotations.py:309]: But aren't `int` and `float` classes too?

Comment (anjali411) [reply torch/jit/annotations.py:309]: ok wait you are right `ann is complex` works!

**Comment by anjali411** [comment]
> Nice! Do you plan to support constant literals in TorchScript code in this diff? If not, I think it is ready to land. If so, I think some more changes are required.

no I'll just make a separate PR for that! also, addressed all the comments and updated the PR

**Review by SplitInfinity — APPROVED** [review]
🚀

**Pull request closed without merging**